CIPs/0016.md

```diff
@@ -0,0 +1,89 @@
+# CIP [0016]: Standard celo protocol URI scheme
+
+- Date: 2020-07-27
+- Author: @yorhodes
+- Status: WIP
+
+## Overview
+
+Create a standard URI scheme for encoding interacting with the Celo protocol for various consumers including SDKs, mobile wallets, and end users.
+Facilitate development of QR systems by standardizing usage and unifying scheme for interoperability.
+
+## Goals
+
+- establish URI standard compatible with mobile app intents that optimizes for **memory** and **performance**
+- enable simple construction of URIs by humans without deep celo protocol understanding
+- enable simple construction of QRs by code from arbitrary celo transactions
+
+## Proposed Solution
+
+### Syntax
+Celo protocol URLs contain "celo" in their schema (protocol) part and are constructed as follows:
+
+    request                 = "celo:" [ prefix "-" ] target [ "/" endpoint ] [ "?" parameters ]
+    prefix                  = "pay" / "request" / "call" / "send"
+    
+    target                  = E164_PHONE_NUMBER / registry_id / celo_address
+    registry_id             = "StableToken" / "GoldToken" / "LockedGold" / ...
+    celo_address            = ( "0x" 40*40HEXDIG )
+    
+    endpoint                = token / function
+    token                   = "cUSD" / "CELO" / ...
+    function                = function_name "(" [ function_args ] ")"
+    function_name           = STRING
+    function_args           = function_arg *( "," function_arg )
+    function_arg            = ( NUMBER / STRING ) [ ":" TYPE ]
+
+    parameters              = parameter *( "&" parameter )
+    parameter               = key "=" value
+    key                     = "value" / "gas" / "gasPrice" / "feeCurrency" / "gatewayFee" / "gatewayFeeRecipient" / "localAmount" / "localCurrencyCode" / ...
+    value                   = NUMBER / STRING / celo_address
+
+### Semantics
+
+`TYPE` is a standard ABI type name, as defined in [Ethereum Contract ABI specification](https://solidity.readthedocs.io/en/develop/abi-spec.html). `STRING` is a URL-encoded unicode string of arbitrary length, where delimiters and the percentage symbol (`%`) are mandatorily hex-encoded with a `%` prefix.
+
+`E164_PHONE_NUMBER` is a standard E164 universal phone number, as defined by the International Telecommunication Union in [E164 Number Specification](https://www.itu.int/rec/T-REC-E.164).
+
+`prefix` is used for differentiating between user and contract `target`s and actions. `pay` and `request` are for user targets and intended for use with phone numbers or addresses and the `token` endpoint. `call` & `send` are for contract targets and intended for use with the registry/addresses and the `function` endpoint. 
```

Comment:
Just an observer's comment, not sure if this makes sense:

Is the use of a dash `-` separator for `prefix` deliberate? It seems to deviate from a typical URI. Would sticking to a slash `/` make more sense, e.g. `celo:prefix/target[/endpoint]`

This would make `prefix` mandatory, and could possibly result in more specific spec for each type of `prefix`.

Reply:
It similar to @carterqw2 's feedback, and keeping a more classic URI, especially one that is similar to existing deeplinks would be a good idea